catchup/networkFetcher.go

```diff
@@ -0,0 +1,114 @@
+				// this is a possible on startup, since the network package might have yet to retrieve the list of peers.
+				time.Sleep(noPeersAvailableSleepInterval)
+			}
+			err = fmt.Errorf("FetchBlock: unable to obtain a list of peers to download the block from : %w", err)
+			return nil, nil, err
+		}
+		peer := psp.Peer
+		httpPeer, ok := peer.(network.HTTPPeer)
+		if ok {
+			return httpPeer, psp, nil
+		}
+		netFetcher.log.Warnf("FetchBlock: non-HTTP peer was provided by the peer selector")
+		netFetcher.peerSelector.rankPeer(psp, peerRankInvalidDownload)
+	}
+	return nil, nil, errors.New("FetchBlock: recurring non-HTTP peer was provided by the peer selector")
+}
+
+// FetchBlock function given a round number returns a block from a http peer
+func (netFetcher *NetworkFetcher) FetchBlock(ctx context.Context, round basics.Round) (*bookkeeping.Block,
+	*agreement.Certificate, time.Duration, error) {
+	// internal retry attempt to fetch the block
+	retryCount = 0
+	for ; retryCount < netFetcher.cfg.CatchupBlockDownloadRetryAttempts; retryCount++ {
+		// keep retrying until a valid http peer is selected by the peerSelector
+		httpPeer, psp, err := netFetcher.getHTTPPeer()
+		if err != nil {
+			return nil, nil, time.Duration(0), err
+		}
+		fetcher := makeUniversalBlockFetcher(netFetcher.log, netFetcher.net, netFetcher.cfg)
+		blk, cert, downloadDuration, err := fetcher.fetchBlock(ctx, round, httpPeer)
+		if err != nil {
+			if ctx.Err() != nil {
+				// caller of the function decided to cancel the download
+				return nil, nil, time.Duration(0), err
+			}
+			netFetcher.log.Infof("FetchBlock: failed to download block %d on attempt %d out of %d. %v", round, retryCount, netFetcher.cfg.CatchupBlockDownloadRetryAttempts, err)
+			netFetcher.peerSelector.rankPeer(psp, peerRankDownloadFailed)
+		} else if !blk.ContentsMatchHeader() && blk.Round() > 0 {
+			netFetcher.peerSelector.rankPeer(psp, peerRankInvalidDownload)
+			// Check if this mismatch is due to an unsupported protocol version
+			if _, ok := config.Consensus[blk.BlockHeader.CurrentProtocol]; !ok {
+				netFetcher.log.Errorf("FetchBlock: downloaded block(%v) unsupported protocol version detected: '%v'", round, blk.BlockHeader.CurrentProtocol)
+			}
+			netFetcher.log.Warnf("FetchBlock: downloaded block(%v) contents do not match header", round)
+			netFetcher.log.Infof("FetchBlock: failed to download block %d on attempt %d out of %d. %v", round, retryCount, netFetcher.cfg.CatchupBlockDownloadRetryAttempts, err)
+		} else {
+			// upon successful download rank the peer according to the download speed
+			peerRank := netFetcher.peerSelector.peerDownloadDurationToRank(psp, downloadDuration)
+			netFetcher.peerSelector.rankPeer(psp, peerRank)
+			return blk, cert, downloadDuration, err
+		}
+	}
+	return nil, nil, time.Duration(0), errors.New("FetchBlock failed after multiple blocks download attempts")
```

Comment:
Will be nice to also report how many attempts were made before it failed.

Reply:
We'll be using only one instance of `NetworkFetcher` that can be shared among multiple goroutines because we need to fetch Peers from the same peerSelector. And as the `fetchBlock` methods makes a GET request its safe to have new instance of `universalBlockFetcher` in every `FetchBlock` right?

Reply:
Is this response posted under the wrong comment?

Reply:
Yes, my bad, for some reason there's no option to directly reply to the comments [https://github.com/algorand/go-algorand/pull/4388#discussion_r945038594](https://github.com/algorand/go-algorand/pull/4388#discussion_r945038594)

Reply:
> We'll be using only one instance of NetworkFetcher that can be shared among multiple goroutines because we need to fetch Peers from the same peerSelector. And as the fetchBlock methods makes a GET request its safe to have new instance of universalBlockFetcher in every FetchBlock right?

I think it is safe. I can't think of why there would be a problem. It is more about code clarity and simplicity. 

But it is important to have a test for that. If you can't finish that test, then you need to create a task ticket for someone else.

Reply:
Sure, I've added the test for concurrent attempts to fetch block.